Split filter related css property parsing out into its own file

weinig opened the pull request:
#### 5984529737faad47906ee14c18c1918f5f5e2f7c
<pre>
Split filter related css property parsing out into its own file
<a href="https://bugs.webkit.org/show_bug.cgi?id=278975">https://bugs.webkit.org/show_bug.cgi?id=278975</a>

Reviewed by Darin Adler.

Split filter related CSS property parsing into its own file, consolidating
parts from CSSPropertyParserHelpers.h/cpp and CSSPropertyParserWorkerSafe.h/cpp.

Took opportunity to clean up refactor parsing and filter operation building
to be easier to read by explicitly handling each filter type. To make things
less error prone, per-function details were abstracted into descriptors
allowing sharing code between the parser and building phases. Additionally,
the builder now follows the pattern of other CSS value conversion code and
asserts that the CSSValue is structured as expected, so it no longer needs
to return a std::optional or be conditionally converted.

Finally, two uses of eager calc() evaluation were removed, moving their
evaluation to the building phase as usual. This is the one change in
behavior, so parsing tests were added and updated.

* LayoutTests/imported/w3c/web-platform-tests/css/filter-effects/parsing/filter-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/filter-effects/parsing/filter-computed.html:
* LayoutTests/imported/w3c/web-platform-tests/css/filter-effects/parsing/filter-parsing-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/filter-effects/parsing/filter-parsing-valid.html:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSFilterFunctionDescriptor.h: Added.
* Source/WebCore/css/CSSFilterImageValue.cpp:
* Source/WebCore/css/CSSFontFeatureValue.cpp:
* Source/WebCore/css/CSSGradientValue.cpp:
* Source/WebCore/css/CSSPrimitiveValue.h:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Filter.cpp: Added.
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Filter.h: Added.
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Font.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Image.cpp:
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
* Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
* Source/WebCore/css/parser/CSSPropertyParserWorkerSafe.cpp: Removed.
* Source/WebCore/css/parser/CSSPropertyParserWorkerSafe.h: Removed.
* Source/WebCore/css/process-css-properties.py:
* Source/WebCore/html/canvas/CanvasRenderingContext2D.cpp:
* Source/WebCore/style/FilterOperationsBuilder.cpp:
* Source/WebCore/style/FilterOperationsBuilder.h:
* Source/WebCore/style/StyleBuilderConverter.h:
* Source/WebCore/style/StyleBuilderState.cpp:
* Source/WebCore/style/StyleBuilderState.h:

Canonical link: <a href="https://commits.webkit.org/283038@main">https://commits.webkit.org/283038@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f273a92e188d93d9a48e54ac5449d7d46de52525

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/65005 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/44372 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/17619 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/69029 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/15611 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/67123 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/52155 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/15893 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/52224 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/10781 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/68071 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/41029 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/56261 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/32846 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/37698 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/13632 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/14487 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/59610 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/13970 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/70734 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/8957 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/13449 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/59552 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/8989 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/56321 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/59781 "23 api tests failed or timed out") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14345 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/7391 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/1066 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/40184 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/41261 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/42442 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/41005 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/css/CSSFontFeatureValue.cpp:
```diff
@@ -46,7 +46,7 @@ String CSSFontFeatureValue::customCSSText() const
     builder.append('"');
     // Omit the value if it's `1` as `1` is implied by default.
-    if (m_value->isCalculated() || m_value->resolveAsIntegerNoConversionDataRequired() != 1)
+    if (auto integerValue = m_value->resolveAsIntegerIfNotCalculated(); !integerValue || *integerValue != 1)
```
Am I remembering correctly that `std::optional` makes `integerValue != 1` do the same thing as `!integerValue || *integerValue != 1`?

weinig replied:
You are! I had no idea that was a thing! That's very nice.

weinig committed:
Split filter related css property parsing out into its own file
https://bugs.webkit.org/show_bug.cgi?id=278975

Reviewed by Darin Adler.

Split filter related CSS property parsing into its own file, consolidating
parts from CSSPropertyParserHelpers.h/cpp and CSSPropertyParserWorkerSafe.h/cpp.

Took opportunity to clean up refactor parsing and filter operation building
to be easier to read by explicitly handling each filter type. To make things
less error prone, per-function details were abstracted into descriptors
allowing sharing code between the parser and building phases. Additionally,
the builder now follows the pattern of other CSS value conversion code and
asserts that the CSSValue is structured as expected, so it no longer needs
to return a std::optional or be conditionally converted.

Finally, two uses of eager calc() evaluation were removed, moving their
evaluation to the building phase as usual. This is the one change in
behavior, so parsing tests were added and updated.

* LayoutTests/imported/w3c/web-platform-tests/css/filter-effects/parsing/filter-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/filter-effects/parsing/filter-computed.html:
* LayoutTests/imported/w3c/web-platform-tests/css/filter-effects/parsing/filter-parsing-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/filter-effects/parsing/filter-parsing-valid.html:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSFilterFunctionDescriptor.h: Added.
* Source/WebCore/css/CSSFilterImageValue.cpp:
* Source/WebCore/css/CSSFontFeatureValue.cpp:
* Source/WebCore/css/CSSGradientValue.cpp:
* Source/WebCore/css/CSSPrimitiveValue.h:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Filter.cpp: Added.
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Filter.h: Added.
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Font.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Image.cpp:
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
* Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
* Source/WebCore/css/parser/CSSPropertyParserWorkerSafe.cpp: Removed.
* Source/WebCore/css/parser/CSSPropertyParserWorkerSafe.h: Removed.
* Source/WebCore/css/process-css-properties.py:
* Source/WebCore/html/canvas/CanvasRenderingContext2D.cpp:
* Source/WebCore/style/FilterOperationsBuilder.cpp:
* Source/WebCore/style/FilterOperationsBuilder.h:
* Source/WebCore/style/StyleBuilderConverter.h:
* Source/WebCore/style/StyleBuilderState.cpp:
* Source/WebCore/style/StyleBuilderState.h:

Canonical link: https://commits.webkit.org/283038@main
